### ISSUE TITLE
SQLite: Enable Decimal

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#19635]
Similar to PR #19617, we can enable more decimal operations by leveraging UDFs. Unlike TimeSpan, however, there isn't an acceptable type we can convert to to perform the operations so it will require roughly one UDF per operation. For example:

.NET | SQL
--- | ---
m1 + m2 | add($m1, $m2)
m1 / m2 | divide($m1, $m2)
m1 > m2 | compare($m1, $m2) > 0
m1 >= m2 | compare($m1, $m2) >= 0
m1 < m2 | compare($m1, $m2) < 0
M1 <= m2 | compare($m1, $m2) <= 0
m1 % m2 | remainder($m1, $m2)
m1 * m2 | multiply($m1, $m2)
m1 - m2 | add($m1, mnegate($m2))
-m | negate($m)
Average(t => t.Decimal) | mavg(t.Decimal)
Max(t => t.Decimal) | mmax(t.Decimal)
Min(t => t.Decimal) | mmin(t.Decimal)
Sum(t => t.Decimal) | msum(t.Decimal)

**Comment by bricelam** [comment]
I don’t think there’s much we can do for OrderBy...

**Comment by lokalmatador** [comment]
Hi, first time here looking to contribute to OSS projects. Being tagged as a good first issue, maybe I can give it a shot?

**Comment by bricelam** [comment]
@lokalmatador Sure! You might want to start small on this one (e.g. just do ef_compare and it’s operators) in your first PR.

**Comment by lokalmatador** [comment]
Great! I guess I start out by looking at your changes in the above mentioned PR ([PR#19617](https://github.com/dotnet/efcore/pull/19617)) to get an idea on how to possibly achieve this. Also, I may ask one or another question the next days :)

**Comment by bricelam** [comment]
Feel free to start a draft PR to ask questions about the implementation

**Comment by bricelam** [comment]
(Added modulo as part of PR #20024)

**Comment by lokalmatador** [comment]
First of, thanks for helping me in getting my first contribution done. I hope I wasn't too much of a burden. Would it be OK to continue on this task? I'd be happy to.

**Comment by bricelam** [comment]
No burden at all! Please feel free to continue

**Comment by lokalmatador** [comment]
Well, then I'd continue with ef_{add, divide, multiply, subtract} and ef_negate.

**Comment by lokalmatador** [comment]
[Draft PR](https://github.com/dotnet/efcore/pull/20212) is here

**Comment by bricelam** [comment]
So, @JonPSmith got me thinking again about ORDER BY and I think we could actually do it via collation.
```cs
connection.CreateCollation(
    "EF_DECIMAL",
    (x, y) => decimal.Compare(decimal.Parse(x), decimal.Parse(y));
```
```sql
SELECT *
FROM MyTable
ORDER BY DecimalColumn COLLATE EF_DECIMAL
```

**Comment by JonPSmith** [comment]
Wow, I wouldn't have thought of that!

I did DM @ErikEJ about this and it put me on to #18593 and I saw the [Sqlite limitations](https://docs.microsoft.com/en-us/ef/core/providers/sqlite/limitations) page.

I use Sqlite in-memory as much as possible for unit testing. That's because a lot of people use 'standard' EF Core, i.e. don't use any of the SQL-specific parts, which means Sqlite in-memory works really well. I used Sqlite in-memory in my last client's project and they had an Azure CI/CD pipeline that includes unit tests and Sqlite in-memory worked really well. 

So, of course, I am interested in differences between Sqlite and say SQL Server (a typical database people use). Other than string case sensitivity and filter/order on decimal, then Sqlite in-memory is a good fit for unit testing. I suspect I will go with a ValueConvertion to double if the database provider is Sqlite to overcome the decimal limitation.

PS. If you guys have any clever way to manage a real database in a CI/CD pipeline then I would love to know. Clearly I could call `EnsureDeleted` and then `EnsureCreated` in every unit test, but that is so slow!

**Comment by ajcvickers** [comment]
@JonPSmith In my experience, it's only the creation and dropping of databases that is really slow. (Also, dropping a database destabilizes SQL Server, so we avoid ever doing it when running tests on the C.I.) We have a [database cleaner](https://github.com/dotnet/efcore/blob/master/test/EFCore.Relational.Specification.Tests/TestUtilities/RelationalDatabaseCleaner.cs) that uses our reverse engineering code to clear the database instead of dropping and re-creating it. We haven't productized it, but we're considering it. Also, it looks like @jbogard's [Respawn](https://github.com/jbogard/Respawn) does something similar.

**Comment by JonPSmith** [comment]
Hi @ajcvickers,

Yes, I have a [cleaner too](https://github.com/JonPSmith/EfCore.TestSupport/blob/master/TestSupport/EfHelpers/WipeDbViaSql.cs) but I think yours might be better, but the real problem is migrations, i.e. the Model changes during development, which is going to happen say 2% (??) of the push to CI/CD pipeline.

I have a fix for handling changes to the Model the that works well locally (and used it in real apps), i.e.
1. Create [unique database names](https://github.com/JonPSmith/EfCore.TestSupport/wiki/4.-Create-SQL-Server-databases#the-createuniqueclassoptions-extension-method) for unit tests, but with a given prefix. 
2. If the Model changes the developer has to manually [run a method](https://github.com/JonPSmith/EfCore.TestSupport/wiki/5.-Quickly-create-empty-database#the-deleteallunittestdatabases-extension-method) that deletes all the databases with names with that prefix.

That works locally, but not for CI/CD. Here are two ideas I am thinking about.
1. At the start of each CI/CD I use my method to delete all the unit test databases before running any unit tests in the pipeline. That simple(ish), but it will be slow for the 98% times there is no Model change. Maybe speed in CI/CD isn't so important(?) 
2. Add a test ahead of the 'delete all unit test databases' that checks if migrations are needed to a 'canary database'. If a migration was applied then it delete all the unit test databases. Quicker in 98%, but feels fragile.

Sorry, I'm unloading my ideas but I want to come up with something for an "article". If its off-topic then please ignore.

**Comment by ajcvickers** [comment]
@JonPSmith The key thing about the cleaner is that it can't be based on the EF model otherwise, as you point out, when the model changes it may no longer match the database structure. That's why the EF one uses the reverse engineering code to figure out what is in the database and then drop it all. The process used in the EF C.I. is that every test database is created if it doesn't already exist, and then cleaned. The database is then never dropped.

**Comment by JonPSmith** [comment]
@ajcvickers Wow - that is so cool! I want it :smile:

Obviously I hadn't looked at your cleaner properly before, but I have now. Its pretty complex so I want to reflect back what I think your cleaner does so you can correct me. I *think* it does the following:
1. If there is an existing database it 
   1. Uses the reverse engineering code to read the schema
   2. The sql code in the cleaner then deletes all the indexes, foreign key (constraints?), tables, and sequences
   3. Then it runs some code given by the caller - for SQL Server this has code to delete other things like views, functions etc.
2. Then it creates the tables etc. defined by the current Model. I presume its uses the code in `context.Database.EnsureCreated`

Is that correct? If so that solves the C.I. problem, but actually makes the local unit testing much better too (no need to run a method when the Model changes). And I tried a few of your unit tests and the initial test is pretty quick too.

The next question is, can the `SqlServerDatabaseCleaner` be made available easily? The `RelationalDatabaseCleaner` doesn't access anything internal, but `SqlServerDatabaseCleaner` does. 

Any comments?

**Comment by jbogard** [comment]
This is exactly what my Respawn tool does, ha!
https://github.com/jbogard/Respawn

**Comment by JonPSmith** [comment]
Hi @jbogard,

I did look at Respawn. The EF Core SQL Server cleaner also deletes views, functions, procs, aggregates, types, schema, but I could copy the EF Core code for that. I notice Respawn is async, which means it can't be used in the unit test constructor, which is a pity. I have a look/think about this.

**Comment by jbogard** [comment]
No for xUnit you need to use its async initialization stuff.

**Comment by JonPSmith** [comment]
OK, looking at that now

**Comment by JonPSmith** [comment]
Hi @jbogard,

I made sure I looked at Respawn and it does a great job of removing all the data, but the EF Core cleaner deletes the actual tables in the database, i.e. Drop(MyTable). Dropping all the database tables, indexes, foreign keys , etc. the what I want, so that unit tests can handle changes in the EF Core Model when running unit tests. 

Also @ajcvickers, looking at Respawn made me think - does your `RelationalDatabaseCleaner` handle circular table references? I assume that deleting the foreign keys first makes that work - correct?

**Comment by jbogard** [comment]
Right, we don’t want that for our integration tests. That’s taken care of
via migrations.

**Comment by ajcvickers** [comment]
Catching up on this. We generally don't use migrations for our functional tests--remember our tests are for testing EF, not an application that uses EF, and hence we have models that change rapidly as we add new features/regression cases to the tests. So what we want is for the database to match the current model, which may be different from any previous models, and hence we re-create the database structure each time. So yes, @JonPSmith, your description seems correct to me.

Dropping database objects is something that has to be done in a specific order to avoid violating constraints--for example, dropping FK constraints before attempting to delete tables that reference the constraint.

The cleaner was written to handle all of the cases that the EF tests generate. It was not generalized beyond this. This means it hasn't been tested against any arbitrary database structure. This is why we haven't productized it--we would need to have confidence (tests) that it would work for at least a large percentage of customer databases. We're considering doing this, or finding some other way of making it usable by others.

**Comment by JonPSmith** [comment]
Thanks @ajcvickers for the update. You said 

> ... hence we have models that change rapidly as we add new features/regression cases to the tests.

This is exactly the situation I, and others, find when running unit tests on an evolving/agile application. In particular I always run unit tests before I build a migration, so I have to handle Model changes before there is a migration. This means your cleaner design is just what I was looking for unit testing, both locally and in a C.I. pipeline. 

I also understand making your current cleaner both work in all states and easy for people to use adds another item to your very full list of changes. Therefore I will try building my own cleaner, maybe only for SQL Server and put it into my [EfCore.TestSupport library](https://github.com/JonPSmith/EfCore.TestSupport). I haven't got time straight straight away, but hopefully I can do something before/when EF Core 5 comes out.

**Comment by ajcvickers** [comment]
@JonPSmith Sounds good. Let us know how it goes.

**Comment by lokalmatador** [comment]
Basic arithmetics ([#20212](https://github.com/dotnet/efcore/pull/20212)) (add, sub, multiply, divide, negate) are done. I would just continue working down the list. I assume for the remaining operations `t` denotes an `IEnumerable`? Is there anything else to be aware of?

**Comment by bricelam** [comment]
Yes, `t` is an `IEnumerable<TEntity>`. I don't have any advice, but @smitpatel might. Aggregates over decimal columns are currently blocked in [SqliteSqlTranslatingExpressionVisitor](https://github.com/dotnet/efcore/blob/master/src/EFCore.Sqlite.Core/Query/Internal/SqliteSqlTranslatingExpressionVisitor.cs). OrderBy is blocked in [SqliteQueryableMethodTranslatingExpressionVisitor](https://github.com/dotnet/efcore/blob/master/src/EFCore.Sqlite.Core/Query/Internal/SqliteQueryableMethodTranslatingExpressionVisitor.cs).

**Comment by smitpatel** [comment]
For aggregate, register the functions an in SqliteSqlTranslatingExpressionVisitor generate translation to that custom SqlFunction.

For order by, now that we have CollateExpression, you can just wrapp the ordering expression in CollateExpression with `EF_DECIMAL` collation (for both TranslateOrderBy/TranslateThenBy).

**Comment by lokalmatador** [comment]
Thanks for the tipps, will give it a shot during this rainy and snowy weekend!

**Comment by bricelam** [comment]
I find it quite the coincidence that D. Richard Hipp [recently added an extension](https://github.com/sqlite/sqlite/blob/master/ext/misc/decimal.c) that does all of this (including the collation)... 🤔

**Comment by JonPSmith** [comment]
Hi @ajcvickers,

I have been using a version of the `EnsureClean` found in the EF Core test code for some time now with excellent results. I'm about to embark on writing the last chapter (hurrah!) of my book which is about unit testing and plan to update my [`EfCore.TestSupport`](https://github.com/JonPSmith/EfCore.TestSupport) open-source libiaray to EF Core 5 and include an `EnsureClean` method which relies on a single class adapted from your `Microsoft.EntityFrameworkCore.TestUtilities.RelationalDatabaseCleaner` class.

My question is what do I need to do to recognize that this code came from the EF Core team? I have currently kept your header with your copyright and added a third line saying I have altered it. Is that enough?
```c#
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
// Altered by Jon P Smith, GitHub @JonPSmith, https://www.thereformedprogrammer.net/
```
Here is a link to the [altered class](https://github.com/JonPSmith/TestSupportSchema/blob/master/TestSupportSchema/Internal/RelationalDatabaseCleaner.cs). 


PS. My library is under an MIT licence.

**Comment by ajcvickers** [comment]
@JonPSmith I'm not a lawyer. :-) @Pilchie?

**Comment by terrajobst** [comment]
IANAL, but MIT requires attribution. We generally handle this by delivering a THIRD-PARTY-NOTICES.txt what we deliver with the apps/packages. See this one [as an example](https://github.com/dotnet/runtime/blob/master/THIRD-PARTY-NOTICES.TXT).

**Comment by JonPSmith** [comment]
Thanks @terrajobst, I have updated the two classes with the following header

```c#
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// Available at https://github.com/dotnet/efcore/blob/main/LICENSE.txt
// Altered by Jon P Smith, GitHub @JonPSmith, https://www.thereformedprogrammer.net/
```

I have also added a section to my project's LICENSE file as follows
```
NOTE

Two classes in this library come from Microsoft's Entity Framework Core
and are therefore Copyright (c) .NET Foundation. All rights reserved.
And licensed under the Apache License, Version 2.0.
The licence is available at https://github.com/dotnet/efcore/blob/main/LICENSE.txt
```

I think that covers it, but if anyone wants me to do anything else I'm happy to oblige.

**Comment by ankurparekh1988** [comment]
Hello, I am looking for my first OSS contribution. Would like to know if this is still open and Can I start looking into this or if you can redirect me on one of open good first issue which has kind of priority ?

**Comment by bricelam** [comment]
@ankurparekh1988 Sure! See my and Smit's comments above for hints on where to start digging into the code. Note, just the aggregate functions are left to do.

**Comment by ankurparekh1988** [comment]
@bricelam @smitpatel  I did try to understand the code and look into Smit's comment as well. However I am bit lost there. Can you please guide me on starting point and I can start from there.

**Comment by dresswithpockets** [comment]
@ankurparekh1988 are you still picking this up? if not, I've already done a bit of work to add aggregate operations and would be able to make a PR soon.

**Comment by ankurparekh1988** [comment]
@dresswithpockets I was working on it but than I got something urgent so I couldn't finish. Since you have already done the changes and can make PR. I would say go ahead. I will look into another open issue. Thanks.

**Comment by dresswithpockets** [comment]
cool.

@bricelam ive got a branch that adds support for the aggregate operations over decimal - how does back porting work? i cant find documentation for how back porting changes works, and this is a change i think would be appropriate to include in 6.x.

**Comment by bricelam** [comment]
We don't backport `enhancement`s. Only very high impact `bug`s and security vulnerabilities. Send a PR against the `main` branch.

**Comment by bricelam** [comment]
But, if *you* want to release something for 6.0, you could create your own library that adds the missing translations. (Something similar to the [EFCore.SqlServer.HierarchyId](https://github.com/efcore/EFCore.SqlServer.HierarchyId) community project.)

**Comment by dresswithpockets** [comment]
Ah, understandable. I suppose I will have to do that, since we use 6.0. I'll send a PR against main soon, thanks!

**Comment by suatsuphi** [comment]
> Sum(t => t.Decimal)	ef_sum(t.Decimal)

hi, can you give an example?
how is it ?

**Comment by bricelam** [comment]
It should look something like this:

```cs
connection.CreateAggregate(
    "ef_sum",
    seed: null,
    (decimal? sum, decimal? value)
        => !value.HasValue
            ? sum
            : !sum.HasValue
                ? value
                : sum + value,
    isDeterministic: true);
```

Careful though, you might run into issue #26070.

**Comment by suatsuphi** [comment]
Hi,
@bricelam thank you.

**Comment by suatsuphi** [comment]
@bricelam 

NotSupportedException: SQLite cannot apply aggregate operator 'Sum' on expressions of type 'decimal'. Convert the values to a supported type, or use LINQ to Objects to aggregate the results on the client side.

I guess this is also necessary for different uses of the sum operator.

**Comment by 21r8390** [comment]
@bricelam @suatsuphi Thank you for catching this bug. 👍 We use SQLite for unit testing and found that all queries containing a decimal do not work. This means that we cannot test the correctness of our queries.

Is there a schedule when this bug will be fixed? Or is there already a workaround?

**Comment by jvakos** [comment]
Any news on this. It's almost 4 years. At least give us a schedule or a workaround.

**Comment by roji** [comment]
@jvakos there's a workaround above in https://github.com/dotnet/efcore/issues/19635#issuecomment-1117563394. Other than that, it has indeed been 4 years, but during which this issue got only 12 votes, meaning that not a lot of users need it.

**Comment by roji** [comment]
> We manly because we use SQLite for unit-testing and until now we are not able to unit-test any functions that sum, max, min or avg decimals in the database.

It's generally discouraged to use a different database for testing (e.g. swap SQL Server for SQLite), since databases - and their providers - have different capabilities; this is an example of such a difference. [See our testing docs](https://learn.microsoft.com/en-us/ef/core/testing/) for more information on this.

**Comment by selangley-wa** [comment]
> > We manly because we use SQLite for unit-testing and until now we are not able to unit-test any functions that sum, max, min or avg decimals in the database.
> 
> It's generally discouraged to use a different database for testing (e.g. swap SQL Server for SQLite), since databases - and their providers - have different capabilities; this is an example of such a difference. [See our testing docs](https://learn.microsoft.com/en-us/ef/core/testing/) for more information on this.

I think most of us here know there are tradeoffs when using a database like SQLite for testing.  If we didn't think some of those tradeoffs still made testing with SQLite a good choice for several scenarios, then we would not be requesting this feature.

**Comment by jvakos** [comment]
Do not think sqlite as a unit testing database. Sqlite is mutch more. We also use sqlite in MAUI apps.

**Comment by njkoopman** [comment]
> Do not think sqlite as a unit testing database. Sqlite is mutch more. We also use sqlite in MAUI apps.

Not having support for decimals made us looking for different options for our MAUI apps. Which I must admit are also not optimal. I rather use SQLite when there is support for decimals.

**Comment by njkoopman** [comment]
> > > We manly because we use SQLite for unit-testing and until now we are not able to unit-test any functions that sum, max, min or avg decimals in the database.
> > 
> > 
> > It's generally discouraged to use a different database for testing (e.g. swap SQL Server for SQLite), since databases - and their providers - have different capabilities; this is an example of such a difference. [See our testing docs](https://learn.microsoft.com/en-us/ef/core/testing/) for more information on this.
> 
> I think most of us here know there are tradeoffs when using a database like SQLite for testing. If we didn't think some of those tradeoffs still made testing with SQLite a good choice for several scenarios, then we would not be requesting this feature.

I agree, this is also true in our case.

> It's generally discouraged to use a different database for testing (e.g. swap SQL Server for SQLite), since databases - and their providers - have different capabilities; this is an example of such a difference. [See our testing docs](https://learn.microsoft.com/en-us/ef/core/testing/) for more information on this.

Although there are some differences between capabilities, using SQLite makes it much easier for us to run test in parallel and faster. A lot of the time these differences are not even relevant for the unit-test. With support for decimals the differences between the data providers would be even less in our case.

**Comment by roji** [comment]
We definitely don't think of SQLite as only a testing database - it's an absolutely important database in many scenarios. However, we do discourage using SQLite as a database fake (test double) when the application actually runs against a different database in production. This used to be much more necessary as testing against the real database used to be difficult, error-prone and slow; but today with containerization technologies, spinning up a containerized database is trivial and fast. I suggest people check out [testcontainers](https://dotnet.testcontainers.org/modules/mssql/), which really makes this trivial.

Beyond that, we do intend to improve decimal support for SQLite, but keep in mind that SQLite itself does not support a decimal data type, or provide any functions for working with them; that's why this isn't trivial. In addition, this issue has only received 13 votes, which means it is generally very low-priority for our users.

**Comment by bjowes** [comment]
We wanted to keep using SQLite with decimals in the DB, since in our test cases the lost precision was not critical. By applying these changes EF Core SQLite will happily convert your decimals to double and support all those missing operators:

In DbContext partial class:
``` csharp
public partial class MyEntities : DbContext 
{
    partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
    {
        if (Database.ProviderName == "Microsoft.EntityFrameworkCore.Sqlite") {
            modelBuilder.Entity<MyTableEntity>(b =>
            {
                b.Property(e => e.OneDecimalProperty).HasConversion<double>();
                b.Property(e => e.AnotherDecimalProperty).HasConversion<double>();
            });
        }
    }
}
```

The condition ensures that the conversions are only applied when the DbContext is used by SQLite, meaning only in our unit tests. 

A drawback is that you need to repeat this for all decimal properties that are included in non-supported operations (such as `.Sum()`, `.Avg()` ...).

**Comment by roji** [comment]
@bjowes usually, the point in using decimal in the first place is to do high-accuracy math; if you're just converting all your decimals to double in the database, it may be better to just have double properties in the first place (i.e. what's the point of making them decimal?).

**Comment by bjowes** [comment]
Agreed @roji - for users with SQLite as their production database, my suggestion would make the behavior similar to using double in the database. However, I think the suggestion is still valuable for others who use SQLite for unit testing (or integration testing) only, and another provider for system tests and production where decimals are properly supported.

I see now that my comment was not clear on that this was the intended scenario.

**Comment by roji** [comment]
@bjowes see my comment just above about it being a bad idea to use Sqlite (or any other database) as a test double. Integration testing - at least in my understanding of the term - would mean that you use Sqlite in production, in which case it again doesn't make much sense to me to use `decimal` but not care about precision.

**Comment by roji** [comment]
> Even with testcontainers, spinning up an SQL Server database with isolation between the tests is still an operation in the order of seconds. That same operation for SQLite in-memory is in the order of milliseconds.

The typical way to use a testcontainer database is indeed not to create a database for each test, but to reuse the same database across tests. That indeed raises the problem of test isolation, but there are well-known ways to deal with that which really aren't that hard ([see our docs for discussion](https://learn.microsoft.com/en-us/ef/core/testing/testing-with-the-database)).

More importantly... Our repo is full of issues from people asking us to, simply put, make the InMemory provider (and the Sqlite provider) behave like SQL Server (or some other database). The pattern is always the same: people start out with the "minimal effort" solution of using InMemory/Sqlite as a database fake, because it's easy, fast and doesn't require thinking about isolation. In all but the most trivial cases, these people run into some place where InMemory/Sqlite simply isn't SQL Server, and their SQL Server queries either don't translate, or return different results. They ask us to make SQLite be case-insensitive with strings (because SQL Server is), implement SQL Server-style full-text search (which is impossible for us to do), or to somehow make the SQL Server-specific functions such as EF.Functions.DateDiff work. None of that is feasible, and it's certainly not the job of the EF SQLite provider to make SQLite behave like SQL Server; SQLite is a database in its own right, with its own behaviors and featureset.

The alternative is to invest a bit of time and effort up-front to handle tes isolation; run all read-only tests against the same database (no isolation problems to begin with), run write tests within a transaction that can be rolled back, and for tests which require exercising code that manages transaction, clean and reseed the database between each test.

> The alternative of mocking out the database completely (repository pattern, mocked context or some other way) is error prone in itself, since it leaves the implementation of acting as a database to the test developer. Detecting changes, resolving and updating references, checking constraints is all on you.

I think there may be a misunderstanding here; the alternative of a repository pattern isn't about mocking out the **database** (that's exactly what Sqlite/InMemory do), but rather mocking out your data layer. In other words, the idea is to cut EF out entirely from your application, and have high-level repository operations which encapsulate stuff like change detection (which is what EF does). That's the right pattern if you want pure unit tests, which verify that **your code** works correctly (and therefore EF shouldn't be involved at all, nor a fake database).

There's unfortunately no middle ground that really works (assuming any but the most trivial cases): you can either do pure unit tests with a repository pattern (fully mocked data access), or you can run integration tests involving your production database system, at which point you need to take care of isolation.

**Comment by selangley-wa** [comment]
In my opinion, one of the dumbest arguments in tech, regardless of the context, is that "My way is better than your way so we are going to do it my way". 

There are always choices, tradeoffs, use cases, even experiments, that will satisfy some need.

If you are a vendor, you can choose your way.  But in the world of open source, everyone is free to do things differently.

On the subject at hand, while I would like to see this request addressed, I perfectly understand the circumstances of a development team that has their own vision of how things should work as well as constraints on time and other resources that require them to make choices on what to support and enhance.

As long as the EF Core Team is willing to accept a reasonable patch or support an extension mechanism by which other parties can integrate their own solution, I see no problem.

Incidentally, there is somewhat of a hack for Sqlite to support a decimal type mentioned here:

NUMERIC and DECIMAL support for Sqlite #2887
https://github.com/launchbadge/sqlx/issues/2887#issuecomment-1820158671

**Comment by njkoopman** [comment]
@bjowes The code that you posted code gave me an idea for our unit-test scenarios without having to mention all the columns specifically.

I put the code below at the end of OnModelCreating:
```cs
            if (Database.ProviderName == "Microsoft.EntityFrameworkCore.Sqlite")
            {
                foreach (Microsoft.EntityFrameworkCore.Metadata.IMutableEntityType entityType in builder.Model.GetEntityTypes())
                {
                    foreach (Microsoft.EntityFrameworkCore.Metadata.IMutableProperty property in entityType.GetProperties())
                    {
                        Type propertyType = Nullable.GetUnderlyingType(property.ClrType) ?? property.ClrType;
                        if (propertyType == typeof(Decimal))
                        {
                            property.SetProviderClrType(typeof(Double));//Decimals are treated as doubles
                        }
                    }
                }
            }
```

This means we cannot test with the complete precision in the unit-tests, but it means we can test all methods using sum, avg, etc. with certain precision, which is a big step forward and maybe even enough for our unit-testing purposes.

Thank you very much for this idea!

**Comment by ranma42** [comment]
Given that https://github.com/dotnet/efcore/issues/26070 has been fixed, I would like to implement the aggregation functions as suggested in https://github.com/dotnet/efcore/issues/19635#issuecomment-1277968681
Is that ok? Is anybody else already working on this?

**Comment by ajcvickers** [comment]
@ranma42 I'll bring it up with the team.

**Comment by ajcvickers** [comment]
@ranma42 We discussed this and we think you should be good to go, but @cincuranet is going to confirm.

**Comment by cincuranet** [comment]
I confirm.

**Comment by erwan-auror** [comment]
Hello, out of interest should we expect the change already merged to be available in 8.0.7 ? Or will those be 9.x only features ?

**Comment by cincuranet** [comment]
It's 9.x feature.

**Comment by JeffreySu** [comment]
@bricelam Does there any updates for this feature?

**Comment by roji** [comment]
@JeffreySu no, otherwise we'd update the table.
